test/features/sync.test.js

```diff
@@ -1,64 +1,695 @@
+
+    console.assert(
+      identities.length === 7,
+      "There should be 7 identities"
+    );
+
+    console.assert(
+      Object.keys(getAssignedSites).length === 5,
+      "There should be 5 site assignments"
+    );
+
+    const personalContainer = 
+      this.syncHelper.lookupIdentityBy(identities, {name: "Personal"});
+    console.log(personalContainer);
+    console.assert(
+      personalContainer.color === "red",
+      "Personal Container should be red"
+    );
+    const mozillaContainer =
+      this.syncHelper.lookupIdentityBy(identities, {name: "Mozilla"});
+    console.assert(
+      mozillaContainer.icon === "pet",
+      "Mozilla Container should be pet"
+    );
+    console.log("!!!Finished!!!");
+  });
+
+  it("CIerrorTest", async function() {
+    await this.syncHelper.stopSyncListeners();
+    console.log("Test state from sync that duped everything initially");
+
+    await this.syncHelper.setState(
+      CI_ERROR_TEST_SYNC, 
+      CI_ERROR_TEST_LOCAL, 
+      CI_ERROR_TEST_IDENTS, 
+      CI_ERROR_TEST_SITES
+    );
+
+    await this.webExt.background.window.sync.runSync();
+
+    const getSync = await this.webExt.browser.storage.sync.get();
+    const getAssignedSites = 
+      await this.webExt.background.window.assignManager.storageArea.getAssignedSites();
+
+    const identities = await this.webExt.browser.contextualIdentities.query({});
+
+    console.assert(
+      Object.keys(getSync.cookieStoreIDmap).length === 7, 
+      "cookieStoreIDmap should have 7 entries"
+    );
+
+    console.assert(
+      identities.length === 7,
+      "There should be 7 identities"
+    );
+
+    console.assert(
+      Object.keys(getAssignedSites).length === 5,
+      "There should be 5 site assignments"
+    );
+
+    const personalContainer = 
+      this.syncHelper.lookupIdentityBy(identities, {name: "Personal"});
+    console.log(personalContainer);
+    console.assert(
+      personalContainer.color === "red",
+      "Personal Container should be red"
+    );
+    const mozillaContainer =
+      this.syncHelper.lookupIdentityBy(identities, {name: "Mozilla"});
+    console.assert(
+      mozillaContainer.icon === "pet",
+      "Mozilla Container should be pet"
+    );
+    console.log("!!!Finished!!!");
+  });
+});
+
+class SyncTestHelper {
+  constructor(webExt) {
+    this.webExt = webExt;
+  }
+
+  async stopSyncListeners() {
+    await this.webExt.browser.storage.onChanged.removeListener(this.webExt.background.window.sync.storageArea.onChangedListener);
+    await this.webExt.background.window.sync.removeContextualIdentityListeners();
+  }
+  
+  async setState(syncData, localData, identityData, assignmentData){
+    await this.removeAllContainers();
+    await this.webExt.browser.storage.sync.clear();
+    await this.webExt.browser.storage.sync.set(syncData);
+    await this.webExt.browser.storage.local.clear();
+    await this.webExt.browser.storage.local.set(localData);
```

Comment:
Awesome. This is what I was having the most trouble with.

Reply:
Thanks for merging. Just realized that I missed some of the comments from the original sync test file, sorry about that.

Reply:
Not sure what you mean, but I'm going over it all.

Reply:
e.g. https://github.com/kendallcorner/multi-account-containers/blob/e6f5bf76c2e501d424a1036a580ca0b16abc3ae4/src/js/background/test.js#L308-L315 is what I meant.

Reply:
> but I'm going over it all.

Great! I'll look into the "browser simulation sync" in the coming days.

If you want to give it a shot, it should work along the lines of:

```js
const webext1 = await initializeWithTab();
const webext2 = await initializeWithTab();

webext1OrigSyncSet = webext1.browser.storage.sync.set;
webext2OrigSyncSet = webext2.browser.storage.sync.set;
webext1OrigSyncRemove = webext1.browser.storage.sync.remove;
webext2OrigSyncRemove = webext2.browser.storage.sync.remove;

webext1.browser.storage.sync.set.callsFake((...args) => {
  webext1OrigSyncSet(...args);
  webext2OrigSyncSet(...args);
  webext2.browser.storage.sync.onChanged.addListener.yield(null, "sync");
});

webext1.browser.storage.sync.remove.callsFake((...args) => {
  webext1OrigSyncRemove(...args);
  webext2OrigSyncRemove(...args);
  webext2.browser.storage.sync.onChanged.addListener.yield(null, "sync");
});

webext2.browser.storage.sync.set.callsFake((...args) => {
  webext2OrigSyncSet(...args);
  webext2OrigSyncSet(...args);
  webext1.browser.storage.sync.onChanged.addListener.yield(null, "sync");
});

webext2.browser.storage.sync.remove.callsFake((...args) => {
  webext2OrigSyncRemove(...args);
  webext2OrigSyncRemove(...args);
  webext1.browser.storage.sync.onChanged.addListener.yield(null, "sync");
});
```

This is a bit verbose, but basically just wires the sync.set&remove calls up between two webextension instances - so calls to e.g.

```js
await webext1.browser.contextualIdentities.create({name: "foo"});
```

should fire the

```js
webext2.browser.storage.onChanged
```

event and hence

```js
await webext2.browser.contextualIdentities.query({});
```

should return the synced container accordingly.